sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/MessageBody.cs

```diff
@@ -142,24 +126,58 @@ public override IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
+                ref IList<ReadOnlyMemory<byte>> refSegments)
             {
-                // fields are lazy initialized to not occupy unnecessary memory when there are no data segments
-                _writer ??= new ArrayBufferWriter<byte>();
-                _segments ??= new List<ReadOnlyMemory<byte>>();
+                int length = 0;
+                int numberOfSegments = 0;
+                List<ReadOnlyMemory<byte>> segments = null;
+                foreach (var segment in dataSegments)
+                {
+                    segments ??= dataSegments is IReadOnlyCollection<Data> readOnlyList
+                        ? new List<ReadOnlyMemory<byte>>(readOnlyList.Count)
+                        : new List<ReadOnlyMemory<byte>>();
+                    ReadOnlyMemory<byte> dataToAppend = segment switch
+                    {
+                        ReadOnlyMemory<byte> romSegment => romSegment,
+                        Data data => data.Value switch
+                        {
+                            byte[] byteArray => byteArray,
+                            ArraySegment<byte> arraySegment => arraySegment,
+                            _ => ReadOnlyMemory<byte>.Empty
+                        },
+                        _ => ThrowArgumentOutOfRange(nameof(segment))
+                    };
+                    length += dataToAppend.Length;
+                    numberOfSegments++;
+                    segments.Add(dataToAppend);
+                }
+
+                if (segments == null)
+                {
+                    return;
+                }
+
+                refWriter = length > 0 ? new ArrayBufferWriter<byte>(length) : new ArrayBufferWriter<byte>();
+                refSegments = segments;
 
-                ReadOnlyMemory<byte> dataToAppend = segment.Value switch
+                for (var i = 0; i < numberOfSegments; i++)
                 {
-                    byte[] byteArray => byteArray,
-                    ArraySegment<byte> arraySegment => arraySegment,
-                    _ => ReadOnlyMemory<byte>.Empty
-                };
-
-                var memory = _writer.GetMemory(dataToAppend.Length);
-                dataToAppend.CopyTo(memory);
-                _writer.Advance(dataToAppend.Length);
-                _segments.Add(memory.Slice(0, dataToAppend.Length));
+                    var dataToAppend = segments[i];
+                    var memory = refWriter.GetMemory(dataToAppend.Length);
+                    dataToAppend.CopyTo(memory);
+                    refWriter.Advance(dataToAppend.Length);
+                    segments[i] = memory.Slice(0, dataToAppend.Length);
+                }
+            }
+
+            private static ReadOnlyMemory<byte> ThrowArgumentOutOfRange(string paramName)
+            {
+                throw new ArgumentOutOfRangeException(paramName, "Provided data segment is not compatible.");
```

Comment:
Should this text be in the `Constants` that have started to replace embedded resources for consistency?

Reply:
I don't think it's necessary since it is only used here. This method could be inlined as well, but also not important.

Reply:
This is essentially a Throw Helper. Originally I wanted to also add the caller and the doesnotreturn attributes but I realized I would need to add them to the sources here which would surface more changes.

Under the assumption this is a throw helper wouldn't the hole point of it be to not online it but use it to increase the chance of the using method to be inlineable due to reduced IL code size? That's why I did not add aggressive inlining here

Reply:
Sorry, I meant that this didn't need to be a separate method, not that we should add the inlining attribute to the method. But will defer to your judgement!

Reply:
Would you prefer if I made this a full blown throw helper including all the bells and whistles? Example https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/ArgumentNullException.cs#L67-L69

Reply:
No, I think it is fine as is or you can just throw directly from the calling method. We already have some helpers in our shared source that you could look at if you like - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/Argument.cs#L190